### ISSUE TITLE
replace call to functions from the string module in twisted.scripts.tap2deb

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5001]
|<img alt="nueces's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| nueces reported|
|-|-|
|Trac ID|trac#5001|
|Type|enhancement|
|Created|2011-03-27 17:30:53Z|
|Branch|https://github.com/twisted/twisted/tree/str-methods-scripts-5001-3|

To make the port to python 3k more easy, I'm replacing all the calls to functions from the string module, for their equivalent from the str type object or calling directly to the built-in method of the string variable.

Attachments:

* [scripts.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b6/7b61dec07e02c188790670af43e717f0f46e8913/e364f7b1c2283319fa6ed748458730e4ab348384.patch) (698 bytes) - added by nueces on 2011-03-27 17:31:07Z - patch
* [tap2deb-5001.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/7b6/7b61dec07e02c188790670af43e717f0f46e8913/20364395faf5ee6efdebf7f0e50d66833d08679c.patch) (6585 bytes) - added by thijs on 2012-07-17 20:51:56Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5001 5001
type__enhancement enhancement
reporter__nueces nueces
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_str_methods_scripts_5001_3 branches/str-methods-scripts-5001-3
branch_author__thijs thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1301247053000000 1301247053000000
changetime__1380571716000000 1380571716000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh<br>@lvh set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_3"></a>|@lvh commented|
|-|-|

(In [31460]) Branching to 'str-methods-scripts-5001'

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_4"></a>|@lvh commented|
|-|-|

This is so trivial and the buildbots are busy checking non-trivial things, I'm not going to wait for a build to complete here.

Added to my merge queue. Thanks as always for your patch :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

This patch modifies untested code.

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_6"></a>|@lvh set owner to nueces<br>@lvh set status to `new`|
|-|-|

What JP said, we should probably get a test here even though it's pretty obviously human-verifiable.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra removed owner|
|-|-|

Attached patch adds tests for tap2deb and replaces the call to string.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

(In [35045]) Branching to 'str-methods-scripts-5001-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_10"></a>|@thijstriemstra commented|
|-|-|

(In [35047]) apply tap2deb-5001.patch and add news file, refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [35053]) replace file with FilePath, fix tests. refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Turns out the test was incorrect and not properly skipped which is fixed now, and `file` was replaced with `FilePath` as well. [buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2 Build results]

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_15"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

9 months ago forgot to correctly put this branch up for review, this is take 2. [Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2).

Also renaming the ticket since string's only used in tap2deb.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_16"></a>|@therve set owner to @thijstriemstra|
|-|-|

1. There are several twistedchecker errors:

```
************* Module twisted.scripts.tap2deb
C0103: 66,4:run: Invalid name "build_dir" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 71,4:run: Invalid name "debian_dir" (should match ([a-z_][a-zA-Z0-9]*)$)
W9402:292,0: The first letter of comment should be capitalized
W9501:199,-1:run: String formatting operations should always use a tuplefor non-mapping values
************* Module twisted.scripts.test.test_tap2deb
W9208: 19,0:TestTap2DEB: Missing docstring
W9208: 23,4:TestTap2DEB.setUp: Missing docstring
W9012: 26,0: Expected 2 blank lines, found 1
W9402: 71,0: The first letter of comment should be capitalized
C0103: 73,8:TestTap2DEB.test_basicOperation: Invalid name "tap_name" (should match ([a-z_][a-zA-Z0-9]*)$)
W9501: 74,33:TestTap2DEB.test_basicOperation: String formatting operations should always use a tuplefor non-mapping values
C0103: 74,8:TestTap2DEB.test_basicOperation: Invalid name "tap_file" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 76,8:TestTap2DEB.test_basicOperation: Invalid name "build_dir" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 77,8:TestTap2DEB.test_basicOperation: Invalid name "input_dir" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 78,8:TestTap2DEB.test_basicOperation: Invalid name "input_name" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 88,8:TestTap2DEB.test_basicOperation: Invalid name "debian_dir" (should match ([a-z_][a-zA-Z0-9]*)$)
```

2. The _checkForDebBuild method is not really useful, you can do the work in setUp directly.

3. Test docstrings generally shouldn't contain "should".

4. As you replaced the os.popen('822-date'), you may as well use the datetime module.

Base on the ticket title I would have been inclined to close this ticket as wontfix, but as a test improvement it looks nice. Please resubmit for review once fixed.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

(In [38445]) address review comments, refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_18"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for your review. Addressed the comments and forced a new [build](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)|@lvh set owner to @lvh|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_20"></a>|@lvh set owner to @thijstriemstra|
|-|-|

Hi thijs,

Looking at the build there seems to be some remaining warts.

pyflakes failures:

```
twisted/scripts/tap2deb.py:59: local variable 'maintainer' is assigned to but never used
twisted/scripts/tap2deb.py:60: local variable 'description' is assigned to but never used
twisted/scripts/tap2deb.py:62: local variable 'long_description' is assigned to but never used
twisted/scripts/tap2deb.py:63: local variable 'twistd_option' is assigned to but never used
twisted/scripts/tap2deb.py:64: local variable 'date' is assigned to but never used
twisted/scripts/tap2deb.py:66: local variable 'python_version' is assigned to but never used
```

twistedchecker errors:

```
************* Module twisted.scripts.tap2deb
W9208:  1,0: Missing docstring
W9208: 15,0:MyOptions: Missing docstring
C0301: 18,0: Line too long (98/79)
C0301: 25,0: Line too long (114/79)
C0103: 38,0: Invalid name "type_dict" (should match (([A-Z_][A-Z0-9_]*)|(__.*__)|([a-z_][a-zA-Z0-9]+)|([A-Z][a-zA-Z0-9]+))$)
W9208: 47,0:run: Missing docstring
C0103: 54,4:run: Invalid name "tap_file" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 55,4:run: Invalid name "base_tap_file" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 57,4:run: Invalid name "deb_file" (should match ([a-z_][a-zA-Z0-9]*)$)
C0301: 60,0: Line too long (87/79)
C0103: 62,4:run: Invalid name "long_description" (should match ([a-z_][a-zA-Z0-9]*)$)
C0103: 63,4:run: Invalid name "twistd_option" (should match ([a-z_][a-zA-Z0-9]*)$)
W9501: 66,21:run: String formatting operations should always use a tuplefor non-mapping values
C0103: 66,4:run: Invalid name "python_version" (should match ([a-z_][a-zA-Z0-9]*)$)
W9401: 96,0: Comments should begin with one whitespace
W9010:120,0: Trailing whitespace found in the end of line
C0301:140,0: Line too long (84/79)
W9501:200,-1:run: String formatting operations should always use a tuplefor non-mapping values
```

Besides the obvious ones (lines too long, missing docstrings...), the pyflakes issues seem to come from using ```% vars()``` instead of specifying interpolation variables explicitly.

I'm hoping that I'm not looking at the wrong file, because e.g.

```
C0103: 66,4:run: Invalid name "python_version" (should match ([a-z_][a-zA-Z0-9]*)$)
```

... was already mentioned by therve in his last review.

cheers and thanks again for working on this

lvh

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_21"></a>|@thijstriemstra commented|
|-|-|

(In [38655]) address review comments, refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_22"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for your review, lvh. I fixed the obvious warnings, but left the names for `vars()` the same to prevent breaking more..

Forced a new [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_23"></a>|@tomprince set owner to @thijstriemstra|
|-|-|

1.At least two of the tests don't need to be skipped, if dpkg-buildpkg isn't installed.
 2. `from email.utils import formatdate as now` ?
   >  As you replaced the os.popen('822-date'), you may as well use the datetime module.

 3. `def run(options=None):`My first guess would be that options would be an instance of `usage.Options`. Perhaps `args` or `argv`? It should also have `@param`.
 4. `self.assertEqual(len(inputDir.listdir()), 4)` seems like something that it would make sense to specify explicitly. And the same for the output (I realize there may be some arch dependence in there, although maybe it should be all?)
 5. There are a couple of changes to the created files. I assume this is due to changes in the underlying tool. Does this mean we don't support older tools now? And if so, what versions.
 6. Since we are using subprocess now, is there any reason to change the directory of the python process?
 7. Since you are touching almost all of the code here, can you update change the variables to match twisted's naming convention.
 8. The docstring for `run` should explain what the script does: it is the script, it doesn't run it.
Please resubmit for review after addressing the above issues.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_24"></a>|@thijstriemstra commented|
|-|-|

(In [38976]) address review comments, refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_25"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for your review.

Replying to [tom.prince](#note_23):
>  1. At least two of the tests don't need to be skipped, if dpkg-buildpkg isn't installed.

Fixed.

>  2. `from email.utils import formatdate as now` ?
>    >  As you replaced the os.popen('822-date'), you may as well use the datetime module.

I could use datetime but it would require more code (unless im missing something):

```python
>>> t = datetime.datetime.utcnow()
>>> now = t.strftime('%a, %d %b %Y %H:%M:%S %z')
'Sat, 29 Jun 2013 10:39:24 '
```

so why use datetime here?

> 
>  3. `def run(options=None):`My first guess would be that options would be an instance of `usage.Options`. Perhaps `args` or `argv`? It should also have `@param`.

Fixed. Also added an additional test for the required `maintainer` argument.

>  4. `self.assertEqual(len(inputDir.listdir()), 4)` seems like something that it would make sense to specify explicitly. And the same for the output (I realize there may be some arch dependence in there, although maybe it should be all?)

Fixed.

>  5. There are a couple of changes to the created files. I assume this is due to changes in the underlying tool. Does this mean we don't support older tools now? And if so, what versions.

I've added the following extra files to the output deb file:

 * `debianDir.child('compat')`. This was to get rid of this warning (and build also fails for me otherwise):
```
dh_clean
dh_clean: No compatibility level specified in debian/compat
dh_clean: This package will soon FTBFS; time to fix it!
```
 * `debianDir.child('source').child('format')`. This seems to be part of a .deb nowadays, but without it tap2deb still functions so I removed it again.

Using `dpkg-buildpackage` on Ubuntu 12.04:
```
$ dpkg-buildpackage --version
Debian dpkg-buildpackage version 1.16.1.2.
```

So without the first extra file the tool doesn't run here (and tests will fail). Does it need a .bugfix file instead?

>  6. Since we are using subprocess now, is there any reason to change the directory of the python process?

`tap2deb` always builds the package in '.build' of the current work directory, so without this change it would build in `_trial_temp` but we want it to build in `_trial_temp/twisted.scripts.test.test_tap2de/TestTap2DEB/test_basicOperation/OruqLn/temp/.build/`.

>  7. Since you are touching almost all of the code here, can you update change the variables to match twisted's naming convention.

Fixed.

>  8. The docstring for `run` should explain what the script does: it is the script, it doesn't run it.

Fixed.


Forced a new [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_26"></a>|@tomprince commented|
|-|-|

>>  6. Since we are using subprocess now, is there any reason to change the directory of the python process?
> `tap2deb` always builds the package in '.build' of the current work directory, so without this change it would build in `_trial_temp` but we want it to build in `_trial_temp/twisted.scripts.test.test_tap2de/TestTap2DEB/test_basicOperation/OruqLn/temp/.build/`.
I was obliquely suggesting changing the working directory of the subprocess without changing the working directory of the trial process.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_27"></a>|@tomprince set owner to @thijstriemstra|
|-|-|

I get the following when trying to run the test on debian squeeze.
```
[FAIL]
Traceback (most recent call last):
  File "/home/tomprince/src/twisted/twisted/scripts/test/test_tap2deb.py", line 91, in test_basicOperation
    ['build-stamp', 'debian', 'install-stamp', 'lemon.tap'])
  File "/home/tomprince/src/twisted/twisted/trial/unittest.py", line 270, in assertEqual
    % (msg, pformat(first), pformat(second)))
twisted.trial.unittest.FailTest: not equal:
a = ['debian', 'lemon.tap']
b = ['build-stamp', 'debian', 'install-stamp', 'lemon.tap']
```

I'd be happier if this ticket didn't try to adjust the behavior of tap2deb to support newer versions of
the dpkg tools, and another ticket was filed to deal with those changes.

Thanks for working on this. Please resumbit for review, after
 1. Reverting the changes to support newer packaging tools.
 2. Changing the tests to not change the directory of the trial process.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_28"></a>|@thijstriemstra commented|
|-|-|

(In [40033]) address review comments, refs [#5001](https://github.com/twisted/twisted/issues/5001)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_29"></a>|@thijstriemstra removed owner|
|-|-|

Thanks for the review.

Replying to [tom.prince](#note_27):
>  1. Reverting the changes to support newer packaging tools.

There was one additional 'compat' change I previously made that is reverted now, tests still pass here.

>  2. Changing the tests to not change the directory of the trial process.

Removed the `os.chdir` calls that were unnecessary.

Forced a [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-2), up for review again.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_30"></a>|@wallrj set owner to @thijstriemstra|
|-|-|

Thanks Thijs,

Notes:
 * Merges cleanly
 * Python3 tests fail. I guess this branch was created before the python3 work.
   1. I created a ticket: https://github.com/twisted-infra/buildslave-config/issues/2
   2. And the python3 tests do pass locally
   3. The dpkg-buildpackage tests pass on Ubuntu
      1. http://buildbot.twistedmatrix.com/builders/natty64-py2.6-epoll/builds/3343/steps/epoll/logs/stdio


Points:
 1. Twistedchecker warning
    1. http://buildbot.twistedmatrix.com/builders/twistedchecker/builds/1305/steps/run-twistedchecker/logs/new%20twistedchecker%20errors
       ```
       ************* Module twisted.scripts.tap2deb
       C0301:130,0: Line too long (86/79)
       ```
    2. There's also another long line in tap2deb which isn't reported by tc.

 2. Python3 syntax errors
    ```
    (Twisted3)[richard@zorin trunk]$ ./admin/run-python3-tests twisted.scripts.test.test_tap2deb
    Traceback (most recent call last):
      File "./admin/run-python3-tests", line 30, in <module>
        unittest.main(module=None, argv=["run-python3-tests", "-v"] + testModules)
      File "/usr/lib64/python3.3/unittest/main.py", line 124, in __init__
        self.parseArgs(argv)
      File "/usr/lib64/python3.3/unittest/main.py", line 168, in parseArgs
        self.createTests()
      File "/usr/lib64/python3.3/unittest/main.py", line 175, in createTests
        self.module)
      File "/usr/lib64/python3.3/unittest/loader.py", line 137, in loadTestsFromNames
        suites = [self.loadTestsFromName(name, module) for name in names]
      File "/usr/lib64/python3.3/unittest/loader.py", line 137, in <listcomp>
        suites = [self.loadTestsFromName(name, module) for name in names]
      File "/usr/lib64/python3.3/unittest/loader.py", line 96, in loadTestsFromName
        module = __import__('.'.join(parts_copy))
      File "/home/richard/projects/Twisted/trunk/twisted/scripts/test/test_tap2deb.py", line 8, in <module>
        from twisted.scripts import tap2deb
      File "/home/richard/projects/Twisted/trunk/twisted/scripts/tap2deb.py", line 163
        debianDir.child('init.d').chmod(0755)
                                           ^
    SyntaxError: invalid token
    ```
    1. Consider fixing that (since you've fixed other Python3 incompatibilities
       in this branch). But on the other hand, the diff is already big, so it
       may not worth fixing that here.

 3. There are one or two bits of ugly formatting / indentation. Consider
    changing those:
    1. ```{or\```
    2. ```% vars()```
    3. ```debianDir.child('README.Debian').setContent(
       '''This package was auto-generated by tap2deb\n''')```
    4. ```error = self.assertRaises(SystemExit, tap2deb.run,
            ["--tapfile", "foo"])```

 4. I searched for other tap2deb tickets and many of the comments seem to
    suggest that it should be abandoned. I agree, I can't imagine using tap2deb
    for deploying twisted applications these days. But I can't actually find a
    deprecation ticket. Shall we create such a ticket?
    1. [#4426](https://github.com/twisted/twisted/issues/4426)#comment:2
    2. [#11235](https://github.com/twisted/twisted/issues/11235)#comment:3
    3. [#11549](https://github.com/twisted/twisted/issues/11549)#comment:5
 5. It seems slightly crazy that we've spent so much time on a tool which is
    probably unused and no longer useful (see irc conversation below)


But anyway, I think the changes look fine and the new test coverage is a good
improvement.

It's been reviewed 6 times by 4 different reviewers.

So please merge after at least fixing the long lines, merging forward and
checking for clean build results.

-RichardW.


Conversation about tap2deb on #twisted-dev

```
14:01 < rwall> And I'm reviewing #5001 which seems like a lot of unnecessary work on an obsolete tool.
14:01 < hynek> rwall: i recall it not being flexible enough for my requirement and IIRC worked under sub-optimal assumptions
14:02 < rwall> Ok, so you might you might have used it. It might be worth maintaining?
14:02 < exarkun> It shouldn't be part of Twisted.
14:02 < hynek> rwall: it’s too long ago but i recall my reaction was more or less “ew” :)
14:03 < rwall> That's what various comments seem to suggest, but I can't find any ticket to actually deprecate it.
14:03 < exarkun> It should be part of Debian.  Since that is probably impossible, it should at least be third-party where it can be kept up to date
                 with Debian Python policy.
14:03 < exarkun> The problem is that any time anyone suggests removing it some random person shows up and says "but!!!"
14:03 < rwall> ok
14:03 < exarkun> (ditto tap2rpm)
14:03 < rwall> yep
14:03 < hynek> loudly deprecate and let rot seems most appropriate.
14:04 < exarkun> I don't want it to rot.  I want it to be a great, working tool.
14:04 < hynek> deploying is too individual IMHO to write a single tool
14:04 < exarkun> It's *already* rotten because no one is willing to maintain it as *part of* Twisted.
14:04 < exarkun> hynek: Who says you have to have a single tool?
14:05 < exarkun> hynek: The solution to "single tool can't do everything" is not "have zero tools" - it's "have many tools".
14:06 < rwall> How does it relate to the debian packaging stuff in setuptools? Isn't that best route?
14:06 < hynek> it would be probably more helpful to properly document how to do it with existing tools like FPM
14:06 < exarkun> rwall: tap2{deb,rpm} are configuration packaging tools
14:07 < exarkun> hynek: If FPM is a pure superset of the functionality, great!  Then tap2{deb,rpm} are redundant and we don't need them.
14:07 < exarkun> I don't know anything about FPM and I'm not likely anytime soon.  Do you think you could write up how FPM replaces these tools?
14:07 < hynek> depending on the aspect, fpm is both a superset and subset :)
14:07 < exarkun> Then I'd be happy to just deprecate them and not bother turning them into third-party projects.
14:08 < hynek> this is not a Twisted problem at all. we’re talking about deploying a python application whose command name is incidentally “twistd”
14:09 < exarkun> hynek: Historically that is an entirely uncompelling argument given the total failure of the Python community to produce deployment
                 tools.
14:09 < exarkun> hynek: I'm happy to be convinced that this is no longer the case - but please don't try to convince me with an IRC conversation.
                 Write something persistent.
14:10 < hynek> to be fair, other communities don’t do better at all. :)
14:10 < exarkun> Or if something is written already, then just give me a link.
14:10 < rwall> exarkun, hynek: ok, I may have opened a can of worms. Thanks for the comments. I'll paste them into the ticket.
14:17 < itamar> all tap2deb does is create a init.d script that starts/stops twistd with some arguments you give it
14:18 < itamar> one hopes that systemd/upstart/whatever make that so easy that having a tool for it is redundant
14:19 < hawkowl> they make it easy, but now you need a tool for making upstart scripts for ubuntu, sysvinit scripts for debian and systemd scripts
                 for god knows what :P
14:21 < rwall> itamar: Oh. So you'd need to create a separate package containing the code anyway.
14:22 < rwall> And come to think of it, the howto spells that out
14:22 < rwall> "Note: tap2rpm and tap2deb do not package your entire application"
14:22 < hynek> well.
14:22 < rwall> https://twistedmatrix.com/documents/current/core/howto/basics.html#auto2
14:23 < hynek> you can use like 5 process managers on debian alone. and it’s fairly trivial nowadays too – the days of SystemV scripts are long
               forgone
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_31"></a>|@thijstriemstra commented|
|-|-|

(In [40139]) Branching to 'str-methods-scripts-5001-3'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_32"></a>|@thijstriemstra set status to `assigned`|
|-|-|

Thanks for the review rwall. I'm going to merge it when the [build](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/str-methods-scripts-5001-3) is green, this ticket escalated already it feels like and if the consensus is to deprecate the tool, then I don't want to waste any more time on it.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_33"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [40141]) Merge str-methods-scripts-5001-3: Replace deprecated string.strip() and add basic tests for twisted.scripts.tap2deb.

Author: thijs, nueces
Reviewer: lvh, exarkun, therve, tom.prince, rwall
Fixes: [#5001](https://github.com/twisted/twisted/issues/5001)
